File: python/trackr/session.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from pathlib import Path
from time import time

from trackr.text_cleaner import clean_track_line, normalize_for_dedupe


def build_session_filename(session_date: date, index: int) -> str:
    if index < 1:
        raise ValueError("session index must be >= 1")
    return f"{session_date.isoformat()}({index})-tracklist.txt"


def choose_next_session_path(output_root: Path, session_date: date | None = None) -> Path:
    day = session_date or date.today()
    output_root.mkdir(parents=True, exist_ok=True)
    index = 1
    while True:
        candidate = output_root / build_session_filename(day, index)
        if not candidate.exists():
            return candidate
        index += 1


def format_elapsed(seconds: int | float) -> str:
    total_seconds = int(max(0, seconds))
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    secs = total_seconds % 60
    if hours > 0:
        return f"{hours}:{minutes:02d}:{secs:02d}"
    return f"{minutes:02d}:{secs:02d}"


@dataclass(frozen=True)
class SessionEntry:
    time: str
    line: str
    rendered: str


class SessionTracker:
    def __init__(self, output_root: Path, timestamps_enabled: bool, delay_seconds: int) -> None:
        self._output_root = output_root
        self._timestamps_enabled = timestamps_enabled
        self._delay_seconds = max(0, delay_seconds)
        self._session_file: Path | None = None
        self._mix_start_at: float | None = None
        self._seen: set[str] = set()

    @property
    def session_file(self) -> Path | None:
        return self._session_file

    def start_new_session(self, session_date: date | None = None) -> Path:
        self._session_file = choose_next_session_path(self._output_root, session_date)
        self._session_file.parent.mkdir(parents=True, exist_ok=True)
        self._session_file.touch(exist_ok=True)
        self._seen.clear()
        self._mix_start_at = None
        self._prime_seen()
        return self._session_file

    def reset_baseline(self) -> None:
        self._mix_start_at = None

    def append(self, line: str, published_at: float | None = None) -> SessionEntry | None:
        if self._session_file is None:
            raise RuntimeError("session is not started")

        cleaned_line = clean_track_line(line)
        if not cleaned_line:
            return None

        normalized = normalize_for_dedupe(cleaned_line)
        if not normalized:
            return None
        if normalized in self._seen:
            return None

        when = time() if published_at is None else published_at
        estimated_track_start = when - self._delay_seconds
        if self._mix_start_at is None:
            self._mix_start_at = estimated_track_start
        rel_seconds = max(0, estimated_track_start - self._mix_start_at)

        timestamp = format_elapsed(rel_seconds) if self._timestamps_enabled else ""
        rendered = f"{timestamp}  {cleaned_line}" if self._timestamps_enabled else cleaned_line

        with self._session_file.open("a", encoding="utf-8", newline="\n") as handle:
            handle.write(f"{rendered}\n")

        self._seen.add(normalized)
        return SessionEntry(time=timestamp, line=cleaned_line, rendered=rendered)

    def _prime_seen(self) -> None:
        if self._session_file is None or not self._session_file.exists():
            return
        for raw in self._session_file.read_text(encoding="utf-8").splitlines():
            normalized = normalize_for_dedupe(raw)
            if normalized:
                self._seen.add(normalized)

File: trackr-electron/src/main/session.ts
/**
 * TRACKR Phase 3C — Session Tracker
 *
 * Port of python/trackr/session.py.
 * Manages per-session tracklist files with deduplication and timestamps.
 */

import { existsSync, mkdirSync, readFileSync, appendFileSync, writeFileSync, unlinkSync } from 'fs';
import { join, dirname } from 'path';
import { cleanTrackLine, normalizeForDedupe } from './cleaner';

export interface SessionEntry {
  time:     string;
  line:     string;
  rendered: string;
}

export function formatElapsed(seconds: number): string {
  const total   = Math.max(0, Math.trunc(seconds));
  const hours   = Math.trunc(total / 3600);
  const minutes = Math.trunc((total % 3600) / 60);
  const secs    = total % 60;
  if (hours > 0) {
    return `${hours}:${String(minutes).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
  }
  return `${String(minutes).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
}

/** Returns "YYYY-MM-DD(N)-tracklist.txt" */
export function buildSessionFilename(date: Date, index: number): string {
  const dateStr = date.toISOString().slice(0, 10); // YYYY-MM-DD
  return `${dateStr}(${index})-tracklist.txt`;
}

/** Finds the next available session file path. Creates outputRoot if needed. */
export function chooseNextSessionPath(outputRoot: string, sessionDate?: Date): string {
  const day = sessionDate ?? new Date();
  mkdirSync(outputRoot, { recursive: true });
  let index = 1;
  while (true) {
    const candidate = join(outputRoot, buildSessionFilename(day, index));
    if (!existsSync(candidate)) return candidate;
    index++;
  }
}

export class SessionTracker {
  private _outputRoot:        string;
  private _timestampsEnabled: boolean;
  private _delaySeconds:      number;
  private _sessionFile:       string | null = null;
  private _mixStartAt:        number | null = null;
  private _seen = new Set<string>();

  constructor(outputRoot: string, timestampsEnabled: boolean, delaySeconds: number) {
    this._outputRoot        = outputRoot;
    this._timestampsEnabled = timestampsEnabled;
    this._delaySeconds      = Math.max(0, delaySeconds);
  }

  get sessionFile(): string | null { return this._sessionFile; }

  startNewSession(sessionDate?: Date): string {
    this._sessionFile = chooseNextSessionPath(this._outputRoot, sessionDate);
    mkdirSync(dirname(this._sessionFile), { recursive: true });
    if (!existsSync(this._sessionFile)) writeFileSync(this._sessionFile, '', 'utf8');
    this._seen.clear();
    this._mixStartAt = null;
    this._primeSeen();
    return this._sessionFile;
  }

  resetBaseline(): void {
    this._mixStartAt = null;
  }

  append(line: string, publishedAt?: number): SessionEntry | null {
    if (!this._sessionFile) throw new Error('session not started');

    const cleanedLine = cleanTrackLine(line);
    if (!cleanedLine) return null;

    const normalized = normalizeForDedupe(cleanedLine);
    if (!normalized) return null;
    if (this._seen.has(normalized)) return null;

    // publishedAt is seconds since epoch (matches Python time.time())
    const when               = publishedAt ?? Date.now() / 1000;
    const estimatedTrackStart = when - this._delaySeconds;
    if (this._mixStartAt === null) this._mixStartAt = estimatedTrackStart;
    const relSeconds = Math.max(0, estimatedTrackStart - this._mixStartAt);

    const timestamp = this._timestampsEnabled ? formatElapsed(relSeconds) : '';
    const rendered  = this._timestampsEnabled ? `${timestamp}  ${cleanedLine}` : cleanedLine;

    appendFileSync(this._sessionFile, `${rendered}\n`, { encoding: 'utf8' });
    this._seen.add(normalized);

    return { time: timestamp, line: cleanedLine, rendered };
  }

  /**
   * Append a suffix (e.g. " [Label, 2024]") to a previously written line.
   * Finds the last line in the session file containing `cleanLine` and appends `suffix`.
   */
  appendSuffix(cleanLine: string, suffix: string): boolean {
    if (!this._sessionFile || !existsSync(this._sessionFile)) return false;
    const content = readFileSync(this._sessionFile, 'utf8');
    const lines = content.split('\n');
    let found = false;
    for (let i = lines.length - 1; i >= 0; i--) {
      if (lines[i].includes(cleanLine) && !lines[i].includes(suffix)) {
        lines[i] = lines[i] + suffix;
        found = true;
        break;
      }
    }
    if (!found) return false;
    writeFileSync(this._sessionFile, lines.join('\n'), 'utf8');
    return true;
  }

  deleteSessionFile(): boolean {
    if (!this._sessionFile || !existsSync(this._sessionFile)) return false;
    unlinkSync(this._sessionFile);
    return true;
  }

  private _primeSeen(): void {
    if (!this._sessionFile || !existsSync(this._sessionFile)) return;
    const content = readFileSync(this._sessionFile, 'utf8');
    for (const raw of content.split('\n')) {
      const normalized = normalizeForDedupe(raw);
      if (normalized) this._seen.add(normalized);
    }
  }
}
